### ISSUE TITLE
No way to specify JSX pragma in config

### ISSUE THREAD

**Issue by foxbunny** [eslint/eslint#9583]
**Tell us about your environment**

* **ESLint Version:** 4.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "browser": true,
    "commonjs": true,
    "es6": true,
    "node": true,
    "jest": true
  },
  "extends": "airbnb-base",
  "plugins": ["import"],
  "settings": {
    "import/resolver": "webpack"
  },
  "parserOptions": {
    "ecmaVersion": 6,
    "ecmaFeatures": {
      "jsx": true,
      "experimentalObjectRestSpread": true
    },
    "sourceType": "module"
  },
  "rules": {
    "jsx-quotes": [
      "error",
      "prefer-double"
    ],
    "arrow-parens": [
      "error",
      "as-needed"
    ],
    "prefer-arrow-callback": [
      "error",
      {
        "allowNamedFunctions": false,
        "allowUnboundThis": false
      }
    ],
    "no-const-assign": "warn",
    "no-this-before-super": "warn",
    "no-undef": "warn",
    "no-unreachable": "warn",
    "no-unused-vars": "warn",
    "no-multiple-empty-lines": [
      "error",
      {
        "max": 2,
        "maxBOF": 1
      }
    ],
    "constructor-super": "warn",
    "semi": [
      "error",
      "never"
    ],
    "quotes": [
      "error",
      "single"
    ],
    "valid-typeof": "warn",
    "comma-dangle": [
      "error",
      {
        "arrays": "always-multiline",
        "objects": "always-multiline",
        "imports": "always-multiline",
        "exports": "always-multiline",
        "functions": "ignore"
      }
    ]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
import * as duckweed from 'duckweed'

const view = ({ model }) => ((_ = model.services.i18n.gettext) => (
  <div>
    {_('Hello world!')}
  </div>
))()
```

**What did you expect to happen?**

There would be a way for me to tell what JSX pragma ESLint should use (`duckweed.html` as configured in Babel).

**What actually happened? Please include the actual, raw output from ESLint.**

There was no (documented) way to specify the pragma, so `dukweed` gets flagged as unused.

```
[eslint] 'duckweed' is defined but never used. (no-unused-vars)
```

**Further explanation**

While it's true that React was the first using JSX, it's not specific to React. In both TypeScript and Babel, it is possible to specify a custom pragma for JSX. There are a number of libraries out there (not just React) that take advantage of this for custom JSX implementations, and even without a library, one can always create custom functions to handle JSX (which is what I do in another project). Furthermore, this pragma is not necessarily specified in the code, as it can be specified in the compiler's configuration (both Babel and TS, for instance).

Considering the above, and considering that ESLint provides support for JSX checking before transpiling, I think it would not be unreasonable to expect JSX pragma config to be available and/or documented.

EDIT: I'm not claiming that this is a widely popular feature, and I fully understand if this is viewed as just another wart that ESLint contributors may end up having to maintain for no good reason. :)

**Comment by platinumazure** [comment]
You can create a [custom rule](https://eslint.org/docs/developer-guide/working-with-rules) or [plugin](https://eslint.org/docs/developer-guide/working-with-plugins) which searches for the imports in question and marks them as used using `context.markVariableAsUsed()`.

The React rule `jsx-uses-vars` takes a similar approach.

**Comment by foxbunny** [comment]
Well, yes, of course I could do that. Point is, JSX is really meaningless without a factory function, so support of JSX is only partial if you cannot specify the function that should be used. Since ESLint provides support for JSX in some form, it would be more complete to include the pragma configuration.

**Comment by ilyavolodin** [comment]
Could you provide a suggestion as to how you think configuration for those should be done? Do you know which rules might be affected by them? Obviously `no-unused-vars`, but any other rules? Should this be an additional configuration on just `no-unused-vars`, or should it be something that is configured in ESLint as a whole?

**Comment by foxbunny** [comment]
The behavior I would expect from the setting is the same as `jsx-uses-vars` rule in the [React plugin](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-uses-react.js) @platinumazure mentions. Not sure if that would affect anything other than `no-unused-vars`. As for how this should be configured, I think a logical place would be in the `parserOptions.ecmaFeatures`:

```
"parserOptions": {
        "ecmaFeatures": {
            "jsx": true | false | string
        }
    },
    ....
}
```

One thing to keep in mind that the factory function may be a property on an object that's currently in the scope. For example, the JSX pragma could be `foo.bar` and the import in the module may be `const foo = require('foo')`. Ideally, ESLint should be able to check for both `foo` and `foo.bar`.

**Comment by foxbunny** [comment]
Sorry for spamming, I actually haven't though through my last comment.

1. There is no need for the setting to be identical to the actual pragma being used, as long as we can specify a name to be checked.
2. ESLint should makr the specified name as being in use *and* check that it's in scope.

**Comment by not-an-aardvark** [comment]
I think this is similar to the `jsx-uses-vars` situation in that it should be handled by a plugin, not by ESLint itself. We don't support react-specific semantics in `no-unused-vars`, so I don't think it makes sense for us to support other custom JSX pragmas.

**Comment by foxbunny** [comment]
> We don't support react-specific semantics, so I don't think it makes sense for us to support other custom JSX pragmas.

I take it this means that beyond just being able to parse JSX, ESLint basically doesn't understand JSX semantics nor it intends to do so? (Not trying to start an argument, just trying to understand the direction so I know how I should invest my time.)

**Comment by not-an-aardvark** [comment]
The term "JSX" describes a syntax extension to JavaScript. However, JSX on its own doesn't have any specific semantics (for example, elements in JSX can have arbitrary names, without any implied relationship to the variables in scope).

Libraries like React use a particular interpretation of JSX syntax and apply special semantics to it. For example, React interprets lowercase element names as strings, and it interprets uppercase element names as variable references. These semantics are part of React, not JSX.

ESLint supports JSX syntax, but it doesn't support the special semantics of libraries that use JSX.

**Comment by foxbunny** [comment]
That notion of JSX conflicts with my own experiments with them. In my experience, transpiler is usually the one that interprets JSX, and factories can only interpret values passed to them by the transpiler. Moreover this seems to be consistent between, say, Babel and TypeScript transpilation, and TypeScript can even pass JSX unmodified to Babel without breaking my code, so at least there is a large consensus about the standard interpretation of the JSX. That's why I assumed that it would be something ESLint would support as well. I guess I misunderstood how it works.

Thanks for your time. You've answered my question so I guess it's time to close this.

**Comment by not-an-aardvark** [comment]
Generally, people use [this babel plugin](https://www.npmjs.com/package/babel-plugin-transform-react-jsx) for JSX when using React, which converts `<div />` into `React.createComponent('div', null)`, and converts `<Component />` into `React.createComponent(Component, null)`. So it is true that the transpiler is interpreting the JSX, but the transpiler is still applying library-specific semantics to do the interpretation.
